### PR TITLE
Prefer scalar type hints over phpdoc annotation

### PR DESCRIPTION
**Pull request by BernhardPosselt**
First stab at fixing #17076 

Still needs testing on PHP 7

@LukasReschke @DeepDiver1975 @MorrisJobke @ChristophWurst @Xenopathic @PVince81

### CODE REVIEW

File: tests/lib/appframework/utility/ControllerMethodReflectorTest.php [hunk 1]
```diff
@@ -104,6 +104,23 @@ public function testReadTypeIntAnnotations(){
 		$this->assertEquals('int', $reader->getType('test'));
 	}
 
+	/**
+	 * @Annotation
+	 * @param int $a
+	 * @requires PHP 7
+	 */
+	public function testReadTypeIntAnnotationsScalarTypes($a, float $b, $c){
```
Comment (Xenopathic) [inline tests/lib/appframework/utility/ControllerMethodReflectorTest.php]: Assign a type annotation for `$b` that isn't `float`, so the unit test checks if scalar type hints are preferred over annotation type hints.

**Comment by Raydiation** [comment]
@Xenopathic good catch, somehow missed the mail in my inbox

please take a look again

**Comment by Xenopathic** [comment]
Looks good. :+1:

Are you going to leave this PR as it is, laying the framework for scalar type hints in the app framework? Or are you going to add more (the WIP tag is still in the description)?

**Comment by Raydiation** [comment]
@Xenopathic nope, that should be all. I don't have a PHP 7 test machine yet, so I'm waiting for someone to run the unit tests on PHP 7 ;)

**Comment by PVince81** [comment]
How to test this ?

**Comment by Xenopathic** [comment]
@PVince81 See the unit tests - if you have a function that looks like that, it will work as the unit tests do.

**Comment by BernhardPosselt** [comment]
Squashed the commits

**Comment by DeepDiver1975** [comment]
:boom: on php7 

````
PHP Warning:  Uncaught PHPUnit_Framework_Error_Warning: Missing argument 1 for OC\AppFramework\Utility\ControllerMethodReflectorTest::testReadTypeIntAnnotationsScalarTypes() in /var/jenkins/workspace/server-master-linux-php7-ci/database/sqlite/label/SLAVE/tests/lib/appframework/utility/ControllerMethodReflectorTest.php:113
11:26:42 Stack trace:
11:26:42 #0 /var/jenkins/workspace/server-master-linux-php7-ci/database/sqlite/label/SLAVE/tests/lib/appframework/utility/ControllerMethodReflectorTest.php(113): PHPUnit_Util_ErrorHandler::handleError(2, 'Missing argumen...', '/var/jenkins/wo...', 113, Array)
11:26:42 #1 [internal function]: OC\AppFramework\Utility\ControllerMethodReflectorTest->testReadTypeIntAnnotationsScalarTypes()
11:26:42 #2 phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php(865): ReflectionMethod->invokeArgs(Object(OC\AppFramework\Utility\ControllerMethodReflectorTest), Array)
11:26:42 #3 phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php(743): PHPUnit_Framework_TestCase->runTest()
11:26:42 #4 phar:///usr/local/bin/phpunit/php-invoker/Invoker. in /var/jenkins/workspace/server-master-linux-php7-ci/database/sqlite/label/SLAVE/tests/lib/appframework/utility/ControllerMethodReflectorTest.php on line 113
11:26:42 PHP Fatal error:  Allowed memory size of 134217728 bytes exhausted (tried to allocate 20480 bytes) in /var/jenkins/workspace/server-master-linux-php7-ci/database/sqlite/label/SLAVE/tests/lib/appframework/utility/ControllerMethodReflectorTest.php on line 113

**Comment by BernhardPosselt** [comment]
Dunno PHP 7 issue? Can't spot the issue in the code

**Comment by Xenopathic** [comment]
@BernhardPosselt It seems that recent PHPUnit versions **require** function parameters to be satisfied by a data provider - add a data provider that just pushes in some dummy data to satisfy it.

**Comment by rullzer** [comment]
Fixed the unit tests.

Time to get this in!
CC: @BernhardPosselt @Xenopathic @PVince81 @DeepDiver1975 @LukasReschke

**Comment by BernhardPosselt** [comment]
Awesome, thanks! :+1:

**Comment by icewind1991** [comment]
:+1: looks good

> I don't have a PHP 7 test machine yet

https://github.com/icewind1991/owncloud-dev/tree/php7 :smile:

**Pull request merged**